Adding unit tests for `useCompositeState` to `Composite` component

andrewhayward opened the pull request:
## What?

This PR adds unit tests to validate `useCompositeState` behaviour.


## Why?

In order to ease [the migration away from `reakit`](https://github.com/WordPress/gutenberg/issues/53278), we need to maintain the current public `Composite` API. `reakit` exposes a `useCompositeState` function, which `ariakit` doesn't directly support, so [we need to port this](https://github.com/WordPress/gutenberg/issues/56548) to [the new `Composite` implementation](https://github.com/WordPress/gutenberg/issues/55224).

To ensure this work maintains existing behaviour, we need to have unit tests in place.


## How?

Tests are written against each prop and action provided by Reakit's `useCompositeState` implementation.

Additionally, tests are written for common use-cases to validate interaction and developer expectation.


## Testing Instructions

- Confirm that test coverage is complete
- Ensure that all tests run

Commit:
Backporting `useCompositeState` for Ariakit

Commit:
Updating CHANGELOG.md

ciampo commented:
Thank you for working on this!

I was wondering if, instead of meticulously testing the hook's arguments and return values, a better approach would be to write tests from the point of view of the consumer of the component — eg

```jsx
const TestExample = () => {
  const state = useCompositeState();

  return (
    <Composite {...state}>
      <CompositeItem>Item 1</CompositeItem>
      <CompositeItem>Item 2</CompositeItem>
      <CompositeItem>Item 3</CompositeItem>
    </Composite>
  );
};

// test the most important features of composite, eg:
// - default roles and semantics
// - default tkeyboard navigation
// - orientation
// - two-dimensional
```

I believe that's a better fit for the needs of this refactor — making sure that we reduce as much as possible the disruption to consumers of the component. What do you think?

diegohaz commented:
Fantastic job! Thanks a lot!

I've got a question, though. Does it make sense to re-implement those `unstable_` props? They're not even a part of the Reakit public API (which means Reakit could have changed them in patch versions).

When these components were re-exported in `@wordpress/components` for internal use, they were not expected to preserve their API forever.

I know practices have shifted. But, with that in mind, I'm wondering if it might be better to provide a minimal API for now, focusing on consumer usage, and only re-implement those obscure features if problems arise.

Commit:
Adding usage tests

Commit:
Adding disabled item tests

Commit:
Adding/updating usage tests

Commit:
Adding shift tests

andrewhayward commented:
> I was wondering if, instead of meticulously testing the hook's arguments and return values, a better approach would be to write tests from the point of view of the consumer...

I've added usage tests to cover the primary modifiers (`orientation`/`loop`/`wrap`/`shift`).

The API tests may not be worth keeping in the long term, if only from an overhead perspective, but I think I'd be more confident including them for now, at least until we're done with the migration.

> I'm wondering if it might be better to provide a minimal API for now, focusing on consumer usage, and only re-implement those obscure features if problems arise.

I think I'd like to leave that conversation for the actual implementation work – as and when we decide that something doesn't need to be included, it'll be easy enough to remove the specific test that goes with it.

Commit:
Tidying

ciampo commented:
I'm personally not sure about the current approach in this PR of testing the exact implementation and signature of the `useCompositeState` component:

As discussed previously, we're aiming for a partial backward compatibility layer that supports the most common way the components are consumed and the most common options. Therefore, I would avoid defining types, and I would not add tests around the hook itself. Instead, I would write tests that simulate how a consumer of the component would use the `useCompositeState` hook together with the rest of the `Composite` components (see previous [suggestion](https://github.com/WordPress/gutenberg/pull/56645#issuecomment-1833757933)).

Potentially, we would then be able to add best-effort backward compatibility by doing something like

```jsx
function useCompositeState(props) {
  const store = useCompositeStore( translateStateArgsToStoreArgs( props ) );
  return { store };
}
```

which should just continue working for consumers adopting the pattern:

```jsx
const state = useCompositeState();

return ( <Composite { ...state }> {/* children */} </Composite> );
```


Although I'd like to hear opinions from other folks too, cc @diegohaz @tyxla @mirka

tyxla commented on packages/components/src/composite/types.ts:
```diff
@@ -0,0 +1,345 @@
+
+type Orientation = 'horizontal' | 'vertical';
+
+type CompositeState = IdState & {
```
Introducing types solely for compatibility tests feels like unnecessary overhead. Especially considering that Ariakit's Composite comes fully typed.

tyxla commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,987 @@
+import type { CompositeInitialState, CompositeStateReturn } from '../types';
+
+const COMPOSITE_SUITES = {
+	reakit: {
```
I do favor the idea of having those thorough tests and being able to test in detail both `Composite` versions with the same suite. I think this can be a good tool to understand the gaps between the 2 implementations.

However, there are a couple of reasons why these tests shouldn't live here in the long term:
* This is testing external functionality. Those tests should live with the canonical underlying library, and not here.
* Let's say we find a gap between v1 and v2. How do we proceed? Will it block the migration until the gap is bridged upstream in `Ariakit`? I think that could potentially be a big unwanted blocker. And yet another pointer that a better place for those tests would be in the underlying library repository. 

So, running those tests against v1 and v2 is a good tool to understand any gaps as part of the migration, but I don't see a good reason to merge the PR and have these tests run all the time. 

With that in mind, I agree with @ciampo's idea of keeping a high-level test suite with general tests from the perspective of the consumer. We don't have to go above and beyond with those IMHO, as it's very easy to step into testing beyond our responsibilities.

Commit:
Removing types and API coverage from tests

andrewhayward commented:
Dropped the typing, removed the full API coverage, and added in support for testing the main ways that Reakit composite components can consume state props.

ciampo commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,536 @@
+				);
+				render( <Test /> );
+
+				await userEvent.tab();
```
We should probably follow the [official docs](https://testing-library.com/docs/user-event/setup/#direct-apis) and call `userEvent.setup()` separately in every test

ciampo commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,536 @@
+			};
+		}
+
+		function useCustomProps( initialState?: InitialState ) {
+			const state = useCompositeState( initialState );
+			const { up, down, previous, next, move } = state;
+
+			return {
+				...state,
+				up: jest.fn( up ),
+				down: jest.fn( down ),
+				previous: jest.fn( previous ),
+				next: jest.fn( next ),
+				move: jest.fn( move ),
+			};
+		}
```
I initially thought that the "custom props" scenario involved passing, as props, the same props that could be passed to the `useCompositeState` hook — if that was the case, implementing backward compatibility would be potentially be as easy as doing

```jsx
const Composite = ( {
  // Destructure all supported "custom props" 
  initialStateProp1,
  initialStateProp2,
  // state and store objects
  state,
  store,
  ...restProps, 
} ) => {
  const internalState = useCompositeState( { initialStateProp1, initialStateProp2 } );

  return (
    <Composite
      store={ store ?? state.store ?? internalState }
      { ...restProps }
    />
  );
}
```

Instead, it looks like those "custom props" are the same as the return value of the `useCompositeState` hook. If that's the case, I'm not sure how we'll be able to support backward-compat for this scenario without writing a lot of custom code, and we may need to drop this scenario in the test

andrewhayward replied:
Yeah, essentially `useCompositeState` returns a collection of functions and values that components can use to read and update the shared composite state.

> These props are returned by the state hook. You can spread them into this component `({...state})` or pass them separately. You can also provide these props from your own state logic.

So there's actually nothing special about what `useCompositeState` returns, beyond it being the easiest way to make sure the state is properly shared. Any or all of the returned props could be created outside of that call.

ciampo replied:
Got it. So, it looks like we wouldn't be able to guarantee backward compatibility on this front, unless we added explicit code supporting each of these props on `Composite` ?

andrewhayward replied:
Yes. Fortunately there aren't that many, so it wouldn't be an insurmountable challenge. "Worth doing?" is of course a different question!

ciampo replied:
I say we don't support this scenario, since we agreed that (at least in the first iteration of adding backwards-compat) we wouldn't add extra code to re-implement reakit functionality, and we would go with a minimal solution instead.

We can probably drop the `useCustomProps` scenario in the tests too, then.

ciampo commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,536 @@
+			};
+		}
+
+		describe.each( [
+			[ 'With "spread" state', useSpreadProps ],
+			[ 'With `state` prop', useStateProps ],
+			[ 'With custom props', useCustomProps ],
+		] )( '%s', ( __, useProps ) => {
```
Neat!

ciampo commented on packages/components/src/composite/test/index.tsx:
Do you think we should also add tests for the `baseId`, `rtl` and `currentId` props? (see [docs](https://reakit.io/docs/composite/#usecompositestate))

andrewhayward replied:
Done.

ciampo reviewed (commented):
Thank you for working on this, @andrewhayward ! I left a couple more comments, but his is definitely moving in the direction that was discussed.

Commit:
Using `userEvent.setup()` rather than `userEvent` directly

Commit:
Adding tests for `baseId`, `currentId` and `rtl`

Commit:
Removing custom prop tests

Commit:
Refactoring, reducing test overhead

ciampo reviewed (approved):
LGTM 🚀 

Let's move forward and iterate on the tests in follow-ups as necessary.
